### PR TITLE
Introducing monetary policy update

### PR DESCRIPTION
**Pull request by AionJayT**
## Notice

It is not allowed to submit your PR to the master branch directly, please submit your PR to the master-pre-merge branch.

## Description

Please include a brief summary of the change that this pull request proposes. Include any relevant motivation and context. List any dependencies required for this change.

- This PR open for reviewing cause it introduces a hard folk. It modifies the current reward calculation to match the description of the Aion white paper. the monetary calculator can be adjustment by initial supply, interest, annum blocks, and the starting blocks. The folk number settings are tentative, might be changed before the release.

  
Fixes Issue # .

## Type of change

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [x] Bug fix.
- [ ] New feature.
- [ ] Enhancement.
- [x] Unit test.
- [ ] Breaking change (a fix or feature that causes existing functionality to not work as expected).
- [ ] Requires documentation update.

## Testing

Please describe the tests you used to validate this pull request. Provide any relevant details for test configurations as well as any instructions to reproduce these results.

-

## Verification

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [x] I have self-reviewed my own code and conformed to the style guidelines of this project.
- [x] New and existing tests pass locally with my changes.
- [x] I have added tests for my fix or feature.
- [x] I have made appropriate changes to the corresponding documentation.
- [x] My code generates no new warnings.
- [x] Any dependent changes have been made.

### CODE REVIEW

**Comment by qoire** [comment]
Whoa! This PR is big, I will be reviewing the MonetaryCalculator and MonetaryCalculatorTest classes.
It would be good to include @SavageLiAion as a reviewer as well if possible.

File: modAion/src/org/aion/zero/api/MonetaryCalculator.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+            CURRENTTERM = term;
+        }
+
+        return CURRENTREWARD;
+    }
+
+    private static BigInteger calculateCompound(long term) {
+        BigInteger compound = DEVIDER.multiply(INITIALSUPPLY);
+        BigInteger preCompound = compound;
+        for (long i = 0; i < term; i++) {
```
Comment (qoire) [inline modAion/src/org/aion/zero/api/MonetaryCalculator.java]: If possible, I would like to bound this function, and assume that any block that violates this bound fails. From there we can generate a LUT for the next 1024 years (so we get O(1)). The reason being that if we were ever to introduce an attack vector that ran this calculation without verifying the block height, a user could reasonably exploit the linear complexity of this calculation to stall the thread executing this.

I will leave final decision up to the implementers. But I would just be wary that this does introduce an avenue of attack.

Comment (qoire) [reply modAion/src/org/aion/zero/api/MonetaryCalculator.java]: Also note: I looked at where rewards are being calculated, I don't believe that there is _currently_ any way to exploit this.

**Comment by AionJayT** [comment]
close this PR, reopen in #852

File: modAion/src/org/aion/zero/api/MonetaryCalculator.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+     * @param blockNum The block number for calculating the reward.
+     * @return The current token reward.
+     * @exception IllegalArgumentException The blockNum should not less or equal to the targeting
+     *     block number.
+     * @exception Exception the method should be called after the Holder initialized.
+     */
+    public static BigInteger getCurrentReward(final long blockNum) {
+        if (Holder.inst == null) {
+            throw new NullPointerException(
+                    "No MonetaryCalculator Holder instance, please initial it before the call.");
```
[the remark this answers is not in the document]
Comment (AionJayT) [reply modAion/src/org/aion/zero/api/MonetaryCalculator.java]: updated in 
https://github.com/aionnetwork/aion/pull/852/commits/2f83196e6d8f13cea399a351c8cf50fb3a77f803

File: modAion/src/org/aion/zero/api/MonetaryCalculator.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        if (term != CURRENTTERM) {
+            CURRENTREWARD = calculateCompound(term);
+            CURRENTTERM = term;
+        }
+
+        return CURRENTREWARD;
+    }
+
+    private static BigInteger calculateCompound(long term) {
+        BigInteger compound = DEVIDER.multiply(INITIALSUPPLY);
```
[the remark this answers is not in the document]
Comment (AionJayT) [reply modAion/src/org/aion/zero/api/MonetaryCalculator.java]: updated in 
https://github.com/aionnetwork/aion/pull/852/commits/2f83196e6d8f13cea399a351c8cf50fb3a77f803

File: modAion/src/org/aion/zero/api/MonetaryCalculator.java (hunk 1 above)
Comment (AionJayT) [reply modAion/src/org/aion/zero/api/MonetaryCalculator.java]: updated in 
https://github.com/aionnetwork/aion/pull/852/commits/2f83196e6d8f13cea399a351c8cf50fb3a77f803

File: modAion/src/org/aion/zero/api/MonetaryCalculator.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+import java.math.BigInteger;
+import java.util.Objects;
+
+/**
+ * @implNote The monetary calculator give the reward base given arguments.
+ * @author Jay Tseng
+ */
+public final class MonetaryCalculator {
+
+    private static BigInteger INITIALSUPPLY;
```
[the remark this answers is not in the document]
Comment (AionJayT) [reply modAion/src/org/aion/zero/api/MonetaryCalculator.java]: updated in 
https://github.com/aionnetwork/aion/pull/852/commits/2f83196e6d8f13cea399a351c8cf50fb3a77f803

File: modAion/src/org/aion/zero/api/MonetaryCalculator.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+                            + " InterestBasePoint: "
+                            + interestBasePoint
+                            + " BlockAnnum: "
+                            + blockAnnum
+                            + " StartingBlockNum:"
+                            + startingBlockNum;
+            throw new IllegalArgumentException(s);
+        }
+
+        INITIALSUPPLY = initialSupply;
```
[the remark this answers is not in the document]
Comment (AionJayT) [reply modAion/src/org/aion/zero/api/MonetaryCalculator.java]: updated in 
https://github.com/aionnetwork/aion/pull/852/commits/2f83196e6d8f13cea399a351c8cf50fb3a77f803

**Pull request closed without merging**